src/core/unit.c

```diff
@@ -1929,6 +1929,102 @@ static bool unit_verify_deps(Unit *u) {
         return true;
 }
 
+static bool unit_can_instantiate(Unit *u, char **ret) {
+        _cleanup_free_ char *t = NULL;
+        unsigned n;
+        int r;
+
+        assert(u);
+        assert(u->manager);
+
+        if (!u->instance)
+                return true;
+
+        if (u->n_max_instances == 0)
+                return true;
+
+        if (!u->manager->instance_counts)
+                return true;
+
+        r = unit_name_template(u->id, &t);
+        if (r < 0)
+                return r;
+
+        n = PTR_TO_UINT(hashmap_get(u->manager->instance_counts, t));
+
+        if (ret)
+                *ret = TAKE_PTR(t);
+        return n < u->n_max_instances;
+}
+
+static int unit_increment_instance_count(Unit *u) {
+        _cleanup_free_ char *t = NULL;
+        unsigned n;
+        int r;
+
+        assert(u);
+        assert(u->manager);
+
+        if (!u->instance)
+                return 0;
+
+        if (u->n_max_instances == 0)
+                return 0;
+
+        r = hashmap_ensure_allocated(&u->manager->instance_counts, &string_hash_ops);
+        if (r < 0)
+                return r;
+
+        r = unit_name_template(u->id, &t);
+        if (r < 0)
+                return r;
+
+        n = PTR_TO_UINT(hashmap_get(u->manager->instance_counts, t));
+        n++;
+
+        if (n == 1)
+                r = hashmap_put(u->manager->instance_counts, t, UINT_TO_PTR(n));
+        else
+                r = hashmap_update(u->manager->instance_counts, t, UINT_TO_PTR(n));
+        if (r < 0)
+                return r;
+
+        if (n == 1)
+                TAKE_PTR(t);
+
+        return 1;
+}
+
+static int unit_decrement_instance_count(Unit *u) {
+        _cleanup_free_ char *t = NULL;
+        unsigned n;
+        int r;
+
+        assert(u);
+        assert(u->manager);
+
+        if (!u->instance)
+                return 0;
+
+        if (u->n_max_instances == 0)
+                return 0;
+
+        r = unit_name_template(u->id, &t);
+        if (r < 0)
+                return r;
+
+        assert(u->manager->instance_counts);
+        n = PTR_TO_UINT(hashmap_get(u->manager->instance_counts, t));
+        assert(n > 0);
```

Comment:
why would this hold? i.e. if the template is not yet in the hashmap this should return NULL, i.e. 0?